diff --git a/src/core/lib/promise/detail/basic_seq.h b/src/core/lib/promise/detail/basic_seq.h
index 663609d419..88bc1582bb 100644
--- a/src/core/lib/promise/detail/basic_seq.h
+++ b/src/core/lib/promise/detail/basic_seq.h
@@ -100,7 +100,7 @@ class BasicSeqIter {
           cur_ = next;
           state_.~State();
           Construct(&state_,
-                    Traits::template CallSeqFactory(f_, *cur_, std::move(arg)));
+                    Traits::CallSeqFactory(f_, *cur_, std::move(arg)));
           return PollNonEmpty();
         });
   }
diff --git a/src/core/lib/surface/server.cc b/src/core/lib/surface/server.cc
index 1b67f69ee0..a9ff1410c1 100644
--- a/src/core/lib/surface/server.cc
+++ b/src/core/lib/surface/server.cc
@@ -93,13 +93,9 @@ struct Server::RegisteredMethod {
   RegisteredMethod(
       const char* method_arg, const char* host_arg,
       grpc_server_register_method_payload_handling payload_handling_arg,
-      uint32_t flags_arg)
-      : method(method_arg == nullptr ? "" : method_arg),
-        host(host_arg == nullptr ? "" : host_arg),
-        payload_handling(payload_handling_arg),
-        flags(flags_arg) {}
+      uint32_t flags_arg);
 
-  ~RegisteredMethod() = default;
+  ~RegisteredMethod();
 
   const std::string method;
   const std::string host;
@@ -200,6 +196,17 @@ class Server::RequestMatcherInterface {
   virtual Server* server() const = 0;
 };
 
+Server::RegisteredMethod::RegisteredMethod(
+    const char* method_arg, const char* host_arg,
+    grpc_server_register_method_payload_handling payload_handling_arg,
+    uint32_t flags_arg)
+    : method(method_arg == nullptr ? "" : method_arg),
+      host(host_arg == nullptr ? "" : host_arg),
+      payload_handling(payload_handling_arg),
+      flags(flags_arg) {}
+
+Server::RegisteredMethod::~RegisteredMethod() = default;
+
 //
 // Server::RequestedCall
 //
diff --git a/src/core/load_balancing/health_check_client.cc b/src/core/load_balancing/health_check_client.cc
index e1317bc04b..1b5b511079 100644
--- a/src/core/load_balancing/health_check_client.cc
+++ b/src/core/load_balancing/health_check_client.cc
@@ -147,20 +147,7 @@ void HealthProducer::HealthChecker::OnConnectivityStateChangeLocked(
   }
 }
 
-void HealthProducer::HealthChecker::StartHealthStreamLocked() {
-  if (GRPC_TRACE_FLAG_ENABLED(grpc_health_check_client_trace)) {
-    gpr_log(GPR_INFO,
-            "HealthProducer %p HealthChecker %p: "
-            "creating HealthClient for \"%s\"",
-            producer_.get(), this,
-            std::string(health_check_service_name_).c_str());
-  }
-  stream_client_ = MakeOrphanable<SubchannelStreamClient>(
-      producer_->connected_subchannel_, producer_->subchannel_->pollset_set(),
-      std::make_unique<HealthStreamEventHandler>(Ref()),
-      GRPC_TRACE_FLAG_ENABLED(grpc_health_check_client_trace) ? "HealthClient"
-                                                              : nullptr);
-}
+  void StartHealthStreamLocked();
 
 void HealthProducer::HealthChecker::NotifyWatchersLocked(
     grpc_connectivity_state state, absl::Status status) {
@@ -317,6 +304,21 @@ class HealthProducer::HealthChecker::HealthStreamEventHandler
   RefCountedPtr<HealthChecker> health_checker_;
 };
 
+void HealthProducer::HealthChecker::StartHealthStreamLocked() {
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_health_check_client_trace)) {
+    gpr_log(GPR_INFO,
+            "HealthProducer %p HealthChecker %p: "
+            "creating HealthClient for \"%s\"",
+            producer_.get(), this,
+            std::string(health_check_service_name_).c_str());
+  }
+  stream_client_ = MakeOrphanable<SubchannelStreamClient>(
+      producer_->connected_subchannel_, producer_->subchannel_->pollset_set(),
+      std::make_unique<HealthStreamEventHandler>(Ref()),
+      GRPC_TRACE_FLAG_ENABLED(grpc_health_check_client_trace) ? "HealthClient"
+                                                              : nullptr);
+}
+
 //
 // HealthProducer::ConnectivityWatcher
 //
